pkg/csrapproval/csrapproval.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package csrapproval handles validation for CSR approval requests.
+package csrapproval
+
+import (
+	"crypto/x509"
+	"fmt"
+	"reflect"
+	"strings"
+
+	authorization "k8s.io/api/authorization/v1beta1"
+	capi "k8s.io/api/certificates/v1beta1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/cloud-provider-gcp/pkg/csrmetrics"
+	"k8s.io/klog"
+	certutil "k8s.io/kubernetes/pkg/apis/certificates/v1beta1"
+)
+
+var nodeClientKeyUsages = []capi.KeyUsage{
+	capi.UsageKeyEncipherment,
+	capi.UsageDigitalSignature,
+	capi.UsageClientAuth,
+}
+
+var nodeServerKeyUsages = []capi.KeyUsage{
+	capi.UsageKeyEncipherment,
+	capi.UsageDigitalSignature,
+	capi.UsageServerAuth,
+}
+
+// Validator represents a workflow to handle a CSR.
+//
+// HandleCSR processes certficate requests
+// according to the decisions made with this interface.
+// See below for details.
+type Validator interface {
+	// Return common parameters for this validator. See definition.
+	Options() Options
+
+	// Should this request be handled by *this* Validator?. Others will be
+	// attempted if you return false here.
+	Recognize(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool
+
+	// If Recognize()'d, then validate the contents of the CSR.
+	// For example, verify that the IP addresses or host names are
+	// permitted by the requestor.
+	Validate(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) (bool, error)
+
+	// Hook function that is called after Validate() is sucessful,
+	// but before final approval. If this function returns an error,
+	// this CSR will be retried.
+	PreApproveHook(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) error
+}
+
+type Options struct {
```

Comment:
Have `Options` implement the `Options` method, so that validator implementations can simply embed it